server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -354,6 +371,22 @@ private static Metadata metadataForSnapshot(SnapshotsInProgress.Entry snapshot,
                     builder.put(indexMetadata, false);
                 }
             }
+
+            if (snapshot.userMetadata() != null && snapshot.userMetadata().containsKey(DataStream.DATA_STREAMS_METADATA_FIELD)) {
+                @SuppressWarnings("unchecked")
+                List<String> dataStreamNames = (List<String>) snapshot.userMetadata().get(DataStream.DATA_STREAMS_METADATA_FIELD);
+                Map<String, DataStream> dataStreams = new HashMap<>();
+                for (String dataStreamName : dataStreamNames) {
+                    DataStream dataStream = metadata.dataStreams().get(dataStreamName);
+                    if (dataStream == null) {
+                        assert snapshot.partial() : "Data stream [" + dataStreamName +
```

Comment:
Another thing I just noticed, we will have to update the list of data streams that gets written to `SnapshotInfo` down the line in case of partial snapshots (if one was concurrently removed) won't we? Currently partial snapshots would list data streams in their `SnapshotInfo` that may not be part of the snapshot because they were concurrently deleted?

Reply:
The list is updated already, `SnapshotInfo` is created by the repository from data streams included in the metadata, not from the list in `Entry`